File: src/main/java/io/github/takenoha/towerdefense/domain/WorldBorderSnapshot.java
package io.github.takenoha.towerdefense.domain;

/** Immutable, Paper-independent projection of a world's square border. */
public record WorldBorderSnapshot(double centerX, double centerZ, double size) {

    public WorldBorderSnapshot {
        requireFinite("centerX", centerX);
        requireFinite("centerZ", centerZ);
        requireFinite("size", size);
        if (size <= 0.0d) {
            throw new IllegalArgumentException("size must be greater than zero");
        }
    }

    /** Returns whether a horizontal circle fits entirely within this border. */
    public boolean containsCircle(double circleCenterX, double circleCenterZ, double radius) {
        requireFinite("circleCenterX", circleCenterX);
        requireFinite("circleCenterZ", circleCenterZ);
        requireFinite("radius", radius);
        if (radius < 0.0d) {
            throw new IllegalArgumentException("radius must not be negative");
        }

        double halfSize = size / 2.0d;
        double minX = centerX - halfSize;
        double maxX = centerX + halfSize;
        double minZ = centerZ - halfSize;
        double maxZ = centerZ + halfSize;
        return circleCenterX - radius >= minX
                && circleCenterX + radius <= maxX
                && circleCenterZ - radius >= minZ
                && circleCenterZ + radius <= maxZ;
    }

    private static void requireFinite(String name, double value) {
        if (!Double.isFinite(value)) {
            throw new IllegalArgumentException(name + " must be finite");
        }
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/domain/WorldBorderSnapshot.kt
package io.github.takenoha.towerdefense.domain

import kotlin.jvm.JvmRecord

/** Immutable, Paper-independent projection of a world's square border. */
@JvmRecord
data class WorldBorderSnapshot(
    val centerX: Double,
    val centerZ: Double,
    val size: Double,
) {
    init {
        requireFinite("centerX", centerX)
        requireFinite("centerZ", centerZ)
        requireFinite("size", size)
        if (size <= 0.0) {
            throw IllegalArgumentException("size must be greater than zero")
        }
    }

    /** Returns whether a horizontal circle fits entirely within this border. */
    fun containsCircle(circleCenterX: Double, circleCenterZ: Double, radius: Double): Boolean {
        requireFinite("circleCenterX", circleCenterX)
        requireFinite("circleCenterZ", circleCenterZ)
        requireFinite("radius", radius)
        if (radius < 0.0) {
            throw IllegalArgumentException("radius must not be negative")
        }

        val halfSize = size / 2.0
        val minX = centerX - halfSize
        val maxX = centerX + halfSize
        val minZ = centerZ - halfSize
        val maxZ = centerZ + halfSize
        return circleCenterX - radius >= minX &&
            circleCenterX + radius <= maxX &&
            circleCenterZ - radius >= minZ &&
            circleCenterZ + radius <= maxZ
    }

    private fun requireFinite(name: String, value: Double) {
        if (!value.isFinite()) {
            throw IllegalArgumentException("$name must be finite")
        }
    }
}
